tests/driver_sht1x/main.c

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (C) 2018 Otto-von-Guericke-Universität Magdeburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup tests
+ * @{
+ *
+ * @file
+ * @brief       Test application for the SHT1X temperature & humidity sensor
+ *
+ * @author      Marian Buschsieweke <marian.buschsieweke@ovgu.de>
+ * @}
+ */
+
+#include <stdio.h>
+
+#include <sht1x.h>
+
+/** @brief Maximum difference from correct temperature value [in e-02 °C] */
+static const int16_t max_diff_temp = 1;
+/** @brief Maximum difference from correct humidity value [in e-02 %] */
+static const int16_t max_diff_hum = 10;
+
+static int16_t expected_temp(const sht1x_dev_t *dev, uint16_t _raw)
+{
+    static const double d1_table[] = {-40.1, -39.8, -39.7, -39.6, - 39.4};
+    double d1 = d1_table[dev->vdd];
+    double d2 = (dev->conf & SHT1X_CONF_LOW_RESOLUTION) ? 0.04 : 0.01;
+    double raw = (double)_raw;
+    double temp = d1 + d2 * raw;
+    return (int16_t)(temp * 100.0);
+}
+
+static int16_t expected_hum(const sht1x_dev_t *dev, uint16_t _raw, int16_t _temp)
+{
+    static const double c1 = -2.0468;
+    static const double t1 = 0.01;
+    double temp = ((double)_temp) / 100.0;
+    double raw = (double)_raw;
+    double c2, c3, t2, hum_linear, hum_real;
+
+    if (dev->conf & SHT1X_CONF_LOW_RESOLUTION){
+        c2 = 0.5872;
+        c3 = -4.0845e-04;
+        t2 = 0.00128;
+    }
+    else {
+        c2 = 0.0367;
+        c3 = -1.5955e-06;
+        t2 = 0.00008;
+    }
+
+    hum_linear = c1 + c2 * raw + c3 * (raw * raw);
+    hum_real = (temp - 25.0) * (t1 + t2 * raw) + hum_linear;
+    return (int16_t)(hum_real * 100.0);
+}
+
+int main(void)
+{
+    const uint8_t vdds[] = {
+        SHT1X_VDD_5_0V, SHT1X_VDD_4_0V, SHT1X_VDD_3_5V, SHT1X_VDD_3_0V,
+        SHT1X_VDD_2_5V,
+    };
+    const uint8_t confs[] = {SHT1X_CONF_LOW_RESOLUTION, 0};
+    const uint16_t max_raw_temps[] = {0xfff, 0x3fff};
+    const uint16_t max_raw_hums[] = {0xff, 0xfff};
+    const char *res_temp[] = {"12", "14"};
+    const char *res_hum[] = {"8", "12"};
+    const char *voltages[] = {"5.0", "4.0", "3.5", "3.0", "2.5"};
+    sht1x_dev_t dev = {.conf = 0};
```

Comment:
Why not use the `sht1x_init` with the default params ? This is the common usage for other sensors.

Reply:
Because this would require an actual sensor to be present, but the unit tests are intended to run on "native" without any hardware. The unit tests verify that the integer arithmetic still gives the same results as the floating point arithmetic given in the data sheet. Because these tests are computationally expensive, it makes no sense to run them on embedded hardware.

Also, using default parameters would dramatically decrease test coverage. The unit tests iterate over all possible configuration states in the for loops and verify that the calculations are correct for all of them - including the configuration that was chosen to be the default one.